src/main/scala/ir/cilvisitor/CILVisitor.scala

```diff
@@ -0,0 +1,196 @@
+package ir.cilvisitor
+
+import ir.*
+import scala.collection.mutable.ArrayBuffer
+
+/** A new visitor based off CIL.
+  *
+  * Copied from ASLi https://github.com/UQ-PAC/aslp/blob/partial_eval/libASL/asl_visitor.ml copying from George Necula's
+  * CIL project (https://people.eecs.berkeley.edu/~necula/cil/)
+  */
+
+enum VisitAction[T]:
+  case SkipChildren()
+  case DoChildren()
+  case ChangeTo(e: T)
+  // changes to e, then visits children of e, then applies f to the result
+  case ChangeDoChildrenPost(e: T, f: T => T)
+
+import VisitAction._
```

Comment:
it would look slightly nicer to me if the enum cases didn't need to be prefixed with VisitAction. no big deal though.